### ISSUE TITLE
Большая сеть

### ISSUE THREAD

**Issue by ufm** [dugdmitry/adhoc_routing#9]
Дмитрий, здравствуйте.
А на какую максимальную (по количеству узлов) сеть расчитан Ваш алгоритм?

**Comment by dugdmitry** [comment]
Добрый день,

максимальное количество узлов зависит от топологии сети (звезда, цепочка, и т.п.), а также пропускной способности соединения, которую Вы хотите поддерживать. Каждый дополнительный узел сети (i.e. router, hop), добавляет дополнительную задержку и сокращает общую пропускную способность маршрута, т.к. он добавляет трафик в сеть, когда пересылает пакет.

В теории, количество узлов на маршруте может быть не ограничено, но пропускная способность такого маршрута будет стремиться к нулю. На практике, данный протокол (RLRP) стабильно работал в WiFi ad-hoc топологии типа "цепочка" с 18 узлами. Кроме того, протокол тестировался и стабильно работал в эмуляторе CORE (https://github.com/coreemu/core) с числом узлов 100+, расставленных в случайной топологии.

Какое количество узлов, топологию или use-case Вы рассматриваете?

**Comment by ufm** [comment]
Я ищу вариант для MESH сети. Узлов, хотя-бы, тысяч на 10. Т.е. не очень много (практически нет) длинных цепочек, но очень сильная связанность. И "носитель" в большинстве случаев - либо ethernet, либо tls тунели между узлами.
Но, как я понимаю, RLRP в данном случае - не лучший вариант.

**Comment by dugdmitry** [comment]
да, RLRP изначально разрабатывался для wireless mesh networks, с использованием raw-socket для работы с беспроводными интерфейсами. Однако он может быть быстро адаптирован для tls туннелей, т.к. написан на Python. Т.е. он потенциально хорошо подходит для целей прототипирования/тестирования L3 уровня.

С точки зрения алгоритма, я не вижу проблем с масштабированием при данном use-case, т.к. пропускная способность не является bottleneck, в отличие от wireless mesh networks. В принципе, любой routing protocol должен работать.

В целом, звучит как интересный use-case. Чтобы адаптировать RLRP, необходимо добавить коммуникацию с tls-туннелем через обычный socket, при условии что есть механизм установления этих туннелей.

**Comment by ufm** [comment]
Менф смущает объём служебного трафика, в случае если сеть становится "большой по объёму".
А еще возникает вопрос о защите от "злонамеренных узлах".
Как минимум:
1. Самый простой - защита от broadcast attack. Когда какой либо узел начинает генерировать поток RREP/RREQ пакетов
2. Защита от узлов, нелегитимно утверждающих, что они "очень хороший путь".
3. Защита от "чёрных дыр". Т.е. от узлов, которые успешно и правильно отвечают о приёме транзитного трафика, но сам трафик просто дропают.

Возможно Вы уже думали на эту тему?

P.S. Для понимания - рассматриваются пути для создания аналога Yggdrasil. Т.е. публичная меш-сеть.

**Comment by dugdmitry** [comment]
да, эти проблемы известны и были частично "приняты во внимание" в текущей имплементации. По поводу объема служебного трафика в связи с большим размером сети - у меня есть идеи как уменьшить количество служебных сообщений или вовсе отказаться от них (zero-control routing), сохранив Reinforcement Learning подход. К сожалению, эти идеи пока не отражены в текущем протоколе, но могут быть достаточно быстро имплементированы.

По поводу механизмов защиты от "злонамеренных узлов" и "broadcast attacks" - в текущей имплементации они не отображены, но они могут быть решены тривиальными методами, на мой взгляд.

Спасибо за референс на Yggdrasil. Судя по всему, там используется свой "link-state" протокол, формирующий и поддерживающий spanning-tree graph. Соответственно, там тоже используются контрольные сообщения для его построения, но их интенсивность скорее всего не зависит от пользовательского трафика. Это имеет свои плюсы и минусы, как и любой подобный link-state (aka proactive) routing протокол.
